src/Utilities/Tags.jl

```diff
@@ -0,0 +1,120 @@
+module Tags
+
+using Compat
+
+const BLOCK_ELEMENTS = Set([
+    :address, :article, :aside, :blockquote, :canvas, :dd, :div, :dl,
+    :fieldset, :figcaption, :figure, :footer, :form, :h1, :h2, :h3, :h4, :h5,
+    :h6, :header, :hgroup, :hr, :li, :main, :nav, :noscript, :ol, :output, :p,
+    :pre, :section, :table, :tfoot, :ul, :video,
+])
+const INLINE_ELEMENTS = Set([
+    :a, :abbr, :acronym, :b, :bdo, :big, :br, :button, :cite, :code, :dfn, :em,
+    :i, :img, :input, :kbd, :label, :map, :object, :q, :samp, :script, :select,
+    :small, :span, :strong, :sub, :sup, :textarea, :time, :tt, :var,
+])
+const VOID_ELEMENTS = Set([
+    :area, :base, :br, :col, :command, :embed, :hr, :img, :input, :keygen,
+    :link, :meta, :param, :source, :track, :wbr,
+])
+const ALL_ELEMENTS = union(BLOCK_ELEMENTS, INLINE_ELEMENTS, VOID_ELEMENTS)
+
+const EMPTY_STRING = ""
+const TEXT = Symbol(EMPTY_STRING)
+
+immutable Tag
+    name :: Symbol
+end
+
+Base.show(io::IO, t::Tag) = print(io, "<", t.name, ">")
+
+macro tags(args...) esc(tags(args)) end
+tags(s) = :(const ($(s...),) = $(map(Tag, s)))
+
+typealias Attributes Vector{Pair{Symbol, String}}
```

Comment:
I'd say that a `Vector` is simpler that a `Dict`. Changing between `Dict` and `Vector` only requires the following change, everything else is generic enough to handle either:

```julia
# typealias Attributes Vector{Pair{Symbol, String}}
typealias Attributes Dict{Symbol, String}
```

A `Dict` has three different `Vector` fields, so it's definitely going to use a bit more space. In this case we don't need to do any attribute access at all, so using a `Dict` really wouldn't be worthwhile. We just need to append results and then print them out in order later.

Switching to a `Dict` gives me the following results:

```
julia> @benchmark div(ul(map(li, s))) # Vector
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  337.30 kb
  allocs estimate:  6020
  minimum time:     140.30 μs (0.00% GC)
  median time:      148.29 μs (0.00% GC)
  mean time:        181.16 μs (16.02% GC)
  maximum time:     2.49 ms (90.70% GC)

julia> @benchmark div(ul(map(li, s))) # Dict
BenchmarkTools.Trial: 
  samples:          7634
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  854.00 kb
  allocs estimate:  9027
  minimum time:     552.36 μs (0.00% GC)
  median time:      568.65 μs (0.00% GC)
  mean time:        654.44 μs (11.50% GC)
  maximum time:     4.84 ms (80.05% GC)
```

Not a *huge* difference, and it does come down slightly once you add attributes. That might have something to do with array resizing and copying though.

(Also, using a `Dict` could give us different orders of attributes in the final rendering on each call  to `julia make.jl` which would be pretty annoying if you want to check whether some code changes cause any subtle changes in the output via `git diff`.)